File: ios/Khrysalis Template/com/lightningkite/khrysalistemplate/vg/FloatingHexagonsDemoVG.shared.swift
//Package: com.lightningkite.khrysalistemplate.vg
//Converted using Khrysalis2

import Foundation
import Khrysalis
import RxSwift
import RxRelay
import KhrysalisMaps



public class FloatingHexagonsDemoVG: ViewGenerator {
    
    
    override public var title: String {
        get {
            return "Floating Hexagons Test"
        }
    }
    
    override public func generate(dependency: ViewDependency) -> View {
        var xml = FloatingHexagonsDemoXml()
        var view = xml.setup(dependency)
        return view
    }
    override public func generate(_ dependency: ViewDependency) -> View {
        return generate(dependency: dependency)
    }
    
    override public init() {
        super.init()
    }
}
 
 

public class FloatingHexagonsDelegate: CustomViewDelegate {
    
    
    public var backgroundPaint: Paint
    public var hexagonPaint: Paint
    public var location: Float? 
    public var moveTo: Float? 
    public var animationTime: Float
    public var hexagons: Array<Hexagon>? 
    
    public func frame(time: Float) -> Void {
        
        if let loc = (location) {
            
            if let moveTo = (moveTo) {
                if abs(loc - moveTo) > Float(0.01), animationTime > 0 {
                    var delta = ( moveTo - loc ) / animationTime
                    hexagons?.forEach{ (it) in 
                        it.moveX(delta)
                    }
                    location = loc + delta
                    animationTime -= time
                } 
            }
 
        }
    }
    public func frame(_ time: Float) -> Void {
        return frame(time: time)
    }
    
    override public func draw(canvas: Canvas, width: Float, height: Float, displayMetrics: DisplayMetrics) -> Void {
        if hexagons == nil {
            var tempList: Array<Hexagon> = []
            
            for i in 0 ... 7 {
                tempList.add(Hexagon(( Random.nextFloat() * width * Float(1.5) ), ( Random.nextFloat() * height ), ( ( Random.nextFloat() + 0.2 ) * width / 4 ).toFloat(), Random.nextInt(1, 5), hexagonPaint))
            }
            hexagons = tempList.toList()
        }
        if location == nil {
            location = width / Float(2)
            moveTo = width / Float(2)
        }
        if backgroundPaint.shader == nil {
            backgroundPaint.shader = LinearGradient(width, Float(0), Float(0), height, [0xFF773DBD.asColor(), 0xFF1226AA.asColor()], [Float(0.25), Float(1)], Shader.TileMode.REPEAT)
        }
        canvas.drawRect(Float(0), Float(0), width, height, backgroundPaint)
        hexagons?.forEach{ (it) in 
            it.draw(canvas)
        }
    }
    override public func draw(_ canvas: Canvas, _ width: Float, _ height: Float, _ displayMetrics: DisplayMetrics) -> Void {
        return draw(canvas: canvas, width: width, height: height, displayMetrics: displayMetrics)
    }
    
    public func moveTo(x: Float, time: Float) -> Void {
        moveTo = x
        animationTime = time
    }
    public func moveTo(_ x: Float, _ time: Float) -> Void {
        return moveTo(x: x, time: time)
    }
    
    override public func onTouchDown(id: Int32, x: Float, y: Float, width: Float, height: Float) -> Bool {
        moveTo(x, Float(10))
        return true
    }
    override public func onTouchDown(_ id: Int32, _ x: Float, _ y: Float, _ width: Float, _ height: Float) -> Bool {
        return onTouchDown(id: id, x: x, y: y, width: width, height: height)
    }
    
    override public func onTouchMove(id: Int32, x: Float, y: Float, width: Float, height: Float) -> Bool {
        return true
    }
    override public func onTouchMove(_ id: Int32, _ x: Float, _ y: Float, _ width: Float, _ height: Float) -> Bool {
        return onTouchMove(id: id, x: x, y: y, width: width, height: height)
    }
    
    override public func onTouchUp(id: Int32, x: Float, y: Float, width: Float, height: Float) -> Bool {
        return true
    }
    override public func onTouchUp(_ id: Int32, _ x: Float, _ y: Float, _ width: Float, _ height: Float) -> Bool {
        return onTouchUp(id: id, x: x, y: y, width: width, height: height)
    }
    
    override public func generateAccessibilityView() -> View?  {
        return nil
    }
    
    override public func sizeThatFitsWidth(width: Float, height: Float) -> Float {
        return height
    }
    override public func sizeThatFitsWidth(_ width: Float, _ height: Float) -> Float {
        return sizeThatFitsWidth(width: width, height: height)
    }
    
    override public func sizeThatFitsHeight(width: Float, height: Float) -> Float {
        return width
    }
    override public func sizeThatFitsHeight(_ width: Float, _ height: Float) -> Float {
        return sizeThatFitsHeight(width: width, height: height)
    }
    
    override public init() {
        let backgroundPaint: Paint = Paint()
        self.backgroundPaint = backgroundPaint
        let hexagonPaint: Paint = Paint()
        self.hexagonPaint = hexagonPaint
        let location: Float?  = nil
        self.location = location
        let moveTo: Float?  = nil
        self.moveTo = moveTo
        let animationTime: Float = Float(0)
        self.animationTime = animationTime
        let hexagons: Array<Hexagon>?  = nil
        self.hexagons = hexagons
        super.init()
        self.hexagonPaint.strokeWidth = Float(4)
        self.hexagonPaint.color = 0xFFFFFFFF.asColor()
        self.backgroundPaint.color = 0xFFFFFFFF.asColor()
        self.backgroundPaint.textSize = Float(12)
        animationFrame.addWeak(self) { (self, time) in 
            self.frame(time)
            self.postInvalidate()
        }
    }
}
 
 
 

public class Hexagon {
    
    public var centerX: Float
    public var centerY: Float
    public var radius: Float
    public var layer: Int32
    public var paint: Paint
    
    
    public func draw(canvas: Canvas) -> Void {
        
        for i in 0.toInt() ... 5.toInt() {
            var angle1: Float = ( i.toFloat() * Float(60) ) * degreesToRadians
            var angle2: Float = ( i.toFloat() * Float(60) + Float(60) ) * degreesToRadians
            var cornerX1 = sin(angle1) * radius + centerX
            var cornerY1 = ( cos(angle1) * Float(1.1) ) * radius + centerY
            var cornerX2 = sin(angle2) * radius + centerX
            var cornerY2 = ( cos(angle2) * Float(1.1) ) * radius + centerY
            paint.alpha = 200 / layer
            canvas.drawLine(cornerX1, cornerY1, cornerX2, cornerY2, paint)
        }
    }
    public func draw(_ canvas: Canvas) -> Void {
        return draw(canvas: canvas)
    }
    
    public func move(x: Float, y: Float) -> Void {
        centerX += x / layer.toFloat()
        centerY += y / layer.toFloat()
    }
    public func move(_ x: Float, _ y: Float) -> Void {
        return move(x: x, y: y)
    }
    
    public func moveX(x: Float) -> Void {
        centerX += x / layer.toFloat()
    }
    public func moveX(_ x: Float) -> Void {
        return moveX(x: x)
    }
    private var degreesToRadians: Float
    
    public init(centerX: Float, centerY: Float, radius: Float, layer: Int32, paint: Paint) {
        self.centerX = centerX
        self.centerY = centerY
        self.radius = radius
        self.layer = layer
        self.paint = paint
        let degreesToRadians: Float = Float(3.14159) / Float(180)
        self.degreesToRadians = degreesToRadians
        print(layer)
    }
    convenience public init(_ centerX: Float, _ centerY: Float, _ radius: Float, _ layer: Int32, _ paint: Paint) {
        self.init(centerX: centerX, centerY: centerY, radius: radius, layer: layer, paint: paint)
    }
}
 

File: android/src/main/java/com/lightningkite/khrysalistemplate/vg/FloatingHexagonsDemoVG.shared.kt
package com.lightningkite.khrysalistemplate.vg

import android.graphics.*
import android.util.DisplayMetrics
import android.view.View
import com.lightningkite.khrysalis.animationFrame
import com.lightningkite.khrysalis.rx.addWeak
import com.lightningkite.khrysalis.views.ViewDependency
import com.lightningkite.khrysalis.views.asColor
import com.lightningkite.khrysalis.views.CustomViewDelegate
import com.lightningkite.khrysalis.views.ViewGenerator
import com.lightningkite.khrysalis.views.draw.newLinearGradient
import com.lightningkite.khrysalistemplate.layouts.FloatingHexagonsDemoXml
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sin
import kotlin.random.Random

class FloatingHexagonsDemoVG() : ViewGenerator() {

    override val title: String
        get() = "Floating Hexagons Test"

    override fun generate(dependency: ViewDependency): View {
        val xml = FloatingHexagonsDemoXml()
        val view = xml.setup(dependency)
        return view
    }
}

class FloatingHexagonsDelegate : CustomViewDelegate() {


    val backgroundPaint: Paint = Paint()
    val hexagonPaint: Paint = Paint()
    init {
        this.hexagonPaint.strokeWidth = 4f
        this.hexagonPaint.color = 0xFFFFFFFF.asColor()
    }
    var location: Float? = null
    var moveTo: Float? = null
    var animationTime: Float = 0f

    var hexagons: List<Hexagon>? = null

    fun frame(time: Float) {
        location?.let { loc ->
            moveTo?.let { moveTo ->
                if (abs(loc - moveTo) > .01f && animationTime > 0) {
                    val delta = (moveTo - loc) / animationTime
                    hexagons?.forEach { it ->
                        it.moveX(delta)
                    }
                    location = loc + delta
                    animationTime -= time
                }
            }
        }
    }

    init {
        this.backgroundPaint.color = 0xFFFFFFFF.asColor()
        this.backgroundPaint.textSize = 12f
        animationFrame.addWeak(this) { self, time ->
            self.frame(time)
            self.postInvalidate()
        }
    }

    override fun draw(canvas: Canvas, width: Float, height: Float, displayMetrics: DisplayMetrics) {
        if (hexagons == null) {
            val tempList: MutableList<Hexagon> = mutableListOf()
            for (i in 0..7) {
                tempList.add(
                    Hexagon(
                        (Random.nextFloat() * width * 1.5f),
                        (Random.nextFloat() * height),
                        ((Random.nextFloat() + 0.2) * width / 4).toFloat(),
                        Random.nextInt(1,5),
                        hexagonPaint
                    )
                )
            }
            hexagons = tempList.toList()
        }
        if (location == null) {
            location = width / 2f
            moveTo = width / 2f
        }
        if (backgroundPaint.shader == null) {
            backgroundPaint.shader = newLinearGradient(
                width,
                0f,
                0f,
                height,
                listOf(0xFF773DBD.asColor(), 0xFF1226AA.asColor()),
                listOf(.25f, 1f),
                Shader.TileMode.REPEAT
            )
        }
        canvas.drawRect(
            0f,
            0f,
            width,
            height,
            backgroundPaint
        )
        hexagons?.forEach { it -> it.draw(canvas) }
    }

    fun moveTo(x: Float, time: Float) {
        moveTo = x
        animationTime = time
    }

    override fun onTouchDown(id: Int, x: Float, y: Float, width: Float, height: Float): Boolean {
        moveTo(x, 10f)
        return true
    }

    override fun onTouchMove(id: Int, x: Float, y: Float, width: Float, height: Float): Boolean {
        return true
    }

    override fun onTouchUp(id: Int, x: Float, y: Float, width: Float, height: Float): Boolean {
        return true
    }


    override fun generateAccessibilityView(): View? = null

    override fun sizeThatFitsWidth(width: Float, height: Float): Float {
        return height
    }

    override fun sizeThatFitsHeight(width: Float, height: Float): Float {
        return width
    }
}


class Hexagon(var centerX: Float, var centerY: Float, val radius: Float, val layer: Int, val paint: Paint) {

    init {
        println(layer)
    }

    fun draw(canvas: Canvas) {
        for (i in 0.toInt()..5.toInt()) {
            val angle1: Float = (i.toFloat() * 60f) * degreesToRadians
            val angle2: Float = (i.toFloat() * 60f + 60f) * degreesToRadians
            val cornerX1 = sin(angle1) * radius + centerX
            val cornerY1 = (cos(angle1) * 1.1f) * radius + centerY
            val cornerX2 = sin(angle2) * radius + centerX
            val cornerY2 = (cos(angle2) * 1.1f) * radius + centerY
            paint.alpha = 200 / layer
            canvas.drawLine(cornerX1, cornerY1, cornerX2, cornerY2, paint)
        }
    }

    fun move(x: Float, y: Float) {
        centerX += x / layer.toFloat()
        centerY += y / layer.toFloat()
    }

    fun moveX(x: Float) {
        centerX += x / layer.toFloat()
    }

    private val degreesToRadians: Float = 3.14159f / 180f
}
